Amortized behaviors of iterators and hasNullReferences should be consistent between WeakHashSet and WeakHashMap

rniwa opened the pull request:
#### 58e8ab36fc7d9b1cb93e6d8b49668546679f5e6a
<pre>
Amortized behaviors of iterators and hasNullReferences should be consistent between WeakHashSet and WeakHashMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=251116">https://bugs.webkit.org/show_bug.cgi?id=251116</a>

Reviewed by Chris Dumez.

Made const_iterator and hasNullReferences of WeakHashSet trigger amortized cleanup as in WeakHashMap for consistency.

* Source/WTF/wtf/WeakHashMap.h:
(WeakHashMap::hasNullReferences): Just increment the operation count and don&apos;t do amortized cleanup here.

* Source/WTF/wtf/WeakHashSet.h:
(WeakHashSet::WeakHashSetConstIterator::WeakHashSetConstIterator):
(WeakHashSet::WeakHashSetConstIterator::operator++): Increment the operation count.
(WeakHashSet::begin):
(WeakHashSet::end):
(WeakHashSet::hasNullReferences): Increase the operation count.
(WeakHashSet::forEach): Don&apos;t increment the operation count here given the iterator does it now.
(WeakHashSet::increaseOperationCountSinceLastCleanup): Now takes the increment as an argument.

* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(WTF_WeakPtr.WeakHashSetAmortizedCleanup): Added.
(WTF_WeakPtr.WeakHashMapAmortizedCleanup):

Canonical link: <a href="https://commits.webkit.org/259341@main">https://commits.webkit.org/259341@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9531a0ca4cada3148efb2bb54e652f0bc0b77a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104646 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13720 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37547 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113921 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174135 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4651 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96976 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112860 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11454 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94483 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39016 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108118 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93310 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26095 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80680 "Passed tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7078 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27453 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92538 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4845 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7185 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4033 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30107 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103475 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47009 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101224 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8974 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3416 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -85,24 +88,16 @@ class WeakHashSet final {
-        increaseOperationCountSinceLastCleanup();
-        return WeakHashSetConstIterator(m_set, m_set.end());
-    }
+    const_iterator begin() const { return WeakHashSetConstIterator(*this, m_set.begin()); }
```
We should probably drop the increaseOperationCountSinceLastCleanup() call in forEach() as well?

cdumez commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -143,7 +138,16 @@ class WeakHashSet final {
+            return !value.get();
+        });
+        if (result)
+            amortizedCleanupIfNeeded(count);
```
I know where doing the same thing in WeakHashMap but this looks surprising to me. The caller is asking hasNullReference() but it may cause up to clean up and drop the null references. This, in itself is surprising. But even more surprising, we'll still return true after we've done the clean up?

rniwa replied:
Maybe shouldn't do cleanup in this function? We can align the behavior of WeakHashMap to WeakHashSet instead.

cdumez replied:
Yes, I'd prefer if we didn't (and aligned WeakHashMap).

cdumez reviewed (approved):
LGTM

rniwa committed:
Amortized behaviors of iterators and hasNullReferences should be consistent between WeakHashSet and WeakHashMap
https://bugs.webkit.org/show_bug.cgi?id=251116

Reviewed by Chris Dumez.

Made const_iterator and hasNullReferences of WeakHashSet trigger amortized cleanup as in WeakHashMap for consistency.

* Source/WTF/wtf/WeakHashMap.h:
(WeakHashMap::hasNullReferences): Just increment the operation count and don't do amortized cleanup here.

* Source/WTF/wtf/WeakHashSet.h:
(WeakHashSet::WeakHashSetConstIterator::WeakHashSetConstIterator):
(WeakHashSet::WeakHashSetConstIterator::operator++): Increment the operation count.
(WeakHashSet::begin):
(WeakHashSet::end):
(WeakHashSet::hasNullReferences): Increase the operation count.
(WeakHashSet::forEach): Don't increment the operation count here given the iterator does it now.
(WeakHashSet::increaseOperationCountSinceLastCleanup): Now takes the increment as an argument.

* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(WTF_WeakPtr.WeakHashSetAmortizedCleanup): Added.
(WTF_WeakPtr.WeakHashMapAmortizedCleanup):

Canonical link: https://commits.webkit.org/259341@main
